src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1169 @@
+    }
+    fn RwLock(comptime T: type) type {
+        return struct { rwlock_ed: T };
+    }
+};
+
+// TODO: move this somewhere better?
+/// NOTE: usage of this type indicates that the integer
+/// must only be operated on using saturating arithmetic.
+/// This is used to annotate types as saturating, to match
+/// agave's type definitions.
+fn Saturating(comptime T: type) type {
+    return enum(T) {
+        _,
+
+        pub fn toInt(self: Saturating(T)) T {
+            return @intFromEnum(self);
+        }
+
+        /// Returns `self`, casted to a pointer to `T`, retaining all attributes.
+        pub fn asInt(self: anytype) AsInt(@TypeOf(self)) {
+            return @ptrCast(self);
+        }
+
+        pub fn AsInt(comptime Self: type) type {
+            const lazy = struct {
+                fn nonMatchErr() noreturn {
+                    @compileError("Expected single item pointer to " ++ @typeName(Saturating(T)));
+                }
+            };
+
+            var info = switch (@typeInfo(Self)) {
+                .Pointer => |info| info,
+                .Optional => |info| switch (@typeInfo(info.child)) {
+                    .Pointer => return ?AsInt(info.child),
+                    else => lazy.nonMatchErr(),
+                },
+                else => lazy.nonMatchErr(),
+            };
+            if (info.size != .One) lazy.nonMatchErr();
+            if (info.child != Saturating(T)) lazy.nonMatchErr();
+            info.child = T;
+
+            return @Type(.{ .Pointer = info });
+        }
+    };
+}
+
+/// AUDIT: https://github.com/anza-xyz/agave/blob/cb32984a9b0d5c2c6f7775bed39b66d3a22e3c46/core/src/consensus/progress_map.rs
+pub const ProgressMap = struct {
+    map: std.AutoArrayHashMapUnmanaged(Slot, ForkProgress),
+
+    pub const INIT: ProgressMap = .{ .map = .{} };
+
+    pub fn deinit(self: ProgressMap, allocator: std.mem.Allocator) void {
+        var map = self.map;
+        for (map.values()) |fork_progress| fork_progress.deinit(allocator);
+        map.deinit(allocator);
+    }
+
+    pub fn clone(
+        self: ProgressMap,
+        allocator: std.mem.Allocator,
+    ) std.mem.Allocator.Error!ProgressMap {
+        var result: ProgressMap = INIT;
+        errdefer result.deinit(allocator);
+
+        try result.map.ensureTotalCapacity(allocator, self.map.count());
+        for (self.map.keys(), self.map.values()) |k, v| {
+            result.map.putAssumeCapacityNoClobber(k, try v.clone(allocator));
+        }
+
+        return result;
+    }
+
+    pub fn getForkStats(self: ProgressMap, slot: Slot) ?*ForkStats {
+        const fork_progress = self.map.getPtr(slot) orelse return null;
+        return &fork_progress.fork_stats;
+    }
+
+    pub fn getHash(self: *const ProgressMap, slot: Slot) ?Hash {
+        const fork_progress = self.map.get(slot) orelse return null;
+        return fork_progress.fork_stats.bank_hash;
+    }
+};
+
+pub const ForkProgress = struct {
```

Comment:
Would you mind adding function(s) to cover the logic in ForkProgress::new and ForkProgress::new_from_bank? For the bank you can stub out the dependency with a basic struct that has the minimal fields necessary.

Reply:
5a91c246b4c3e8f1ee5dfd062f5edba8280c6ed4
52b42b82333ca9ebd4cca7bc534cefb46f43e532